src/Pact/Eval.hs

```diff
@@ -1226,15 +1228,85 @@ reduceDirect r _ ai = evalError ai $ "Unexpected non-native direct ref: " <> pre
 
 initPact :: Info -> PactContinuation -> Term Ref -> Eval e (Term Name)
 initPact i app bod = view eePactStep >>= \es -> case es of
-  Just v -> evalError i $ "initPact: internal error: step already in environment: " <> pretty v
-  Nothing -> view eeHash >>= \hsh ->
-    applyPact i app bod $ PactStep 0 False (toPactId hsh) Nothing
+  Just v@(PactStep step b (PactId parent) _) -> do
+    whenExecutionFlagSet FlagDisablePact43 $
+      evalError i $ "initPact: internal error: step already in environment: " <> pretty v
+    let Hash name' = pactHash $ T.encodeUtf8 $ renderCompactText (_pcDef app)
+        newPactId = toPactId (pactHash (T.encodeUtf8 parent <> ":" <> name'))
+    applyNestedPact i app bod $ PactStep step b newPactId Nothing
+  Nothing -> view eeHash >>= \hsh -> do
+    let pStep = PactStep 0 False (toPactId hsh) Nothing
+    local (set eePactStep (Just pStep)) $ applyPact i app bod pStep mempty
+
+-- Nested defpact application, subject to a few conditions:
+-- - If parent step has rollback, this step _Also_ needs to have rollback.
+applyNestedPact :: Info -> PactContinuation -> Term Ref -> PactStep -> Eval e (Term Name)
+applyNestedPact i app (TList steps _a _b) ps@PactStep {..} = do
+  -- only one pact state allowed in a transaction
+  parentExec <- use evalPactExec >>= \case
+    Nothing -> evalError i $ "Nested Pact attempted but no pactExec found for:" <> pretty (_pcDef app)
+    -- Nested pact execution
+    Just pe -> pure pe
+
+  -- retrieve indicated step from code
+  st <- maybe (evalError i $ "applyPact: step not found: " <> pretty _psStep) return $ steps V.!? _psStep
+  step <- case st of
+    TStep step _meta _i -> return step
+    t -> evalError (_tInfo t) "expected step"
 
+  -- determine if step is skipped (for private execution)
```

Comment:
Let's just disallow nested for private pacts altogether. The `entity` designation doesn't really extend to this concept.

Reply:
:+1: